test/xebow_test.exs

```diff
@@ -3,31 +3,223 @@ defmodule XebowTest do
+      end
+    end
+  end
+
+  setup %{single_animation: single_animation} do
+    Xebow.set_active_animation_types(single_animation)
+  end
+
+  describe "can get the config and schema" do
+    test "of the current active animation", %{
+      single_animation: single_animation,
+      single_schema: single_schema
+    } do
+      [animation_module] = single_animation
+      config_module = Module.concat(animation_module, "Config")
+
+      assert {%^config_module{}, ^single_schema} = Xebow.get_animation_config()
+    end
+
+    test "or nil when no animations active" do
+      Xebow.set_active_animation_types([])
+
+      assert Xebow.get_animation_config() == nil
+    end
+  end
+
+  test "config of the current animation can be updated", %{} do
+    single_config = Xebow.get_animation_config()
+    update_params = %{test_field: :b}
+
+    assert Xebow.update_animation_config(update_params) == :ok
+    refute Xebow.get_animation_config() == single_config
+
+    {config, _schema} = Xebow.get_animation_config()
+
+    assert config.test_field == :b
   end
 
-  test "can get and set active animation types" do
-    animation_types = [MockAnimations.Type1]
+  describe "configurables" do
+    setup [:create_config_fn, :create_unregister_config_fn]
+
+    test "can be registered, which is idempotent", %{
+      config_fn: config_fn
+    } do
+      assert Xebow.register_configurable(config_fn) == {:ok, config_fn}
+      assert Xebow.register_configurable(config_fn) == {:ok, config_fn}
 
-    assert Xebow.get_active_animation_types() != animation_types
+      Xebow.next_animation()
+      config = Xebow.get_animation_config()
+
+      assert_receive {:config, ^config}
+      refute_receive {:config, ^config}
+    end
 
-    Xebow.set_active_animation_types(animation_types)
+    test "can be unregistered, which is idempotent", %{
+      config_fn: config_fn
+    } do
+      Xebow.register_configurable(config_fn)
+
+      assert Xebow.unregister_configurable(config_fn) == :ok
+      assert Xebow.unregister_configurable(config_fn) == :ok
+
+      Xebow.previous_animation()
+      config = Xebow.get_animation_config()
+
+      refute_receive {:config, ^config}
+    end
 
-    assert Xebow.get_active_animation_types() == animation_types
+    test "are called when switching animations", %{
+      config_fn: config_fn
+    } do
+      Xebow.register_configurable(config_fn)
+
+      Xebow.next_animation()
+      config = Xebow.get_animation_config()
+      assert_receive {:config, ^config}
+    end
+
+    test "can return :unregister to unregister themselves", %{
+      unregister_config_fn: unregister_config_fn,
+      unregister_message: unregister_message
+    } do
+      assert Xebow.register_configurable(unregister_config_fn) == {:ok, unregister_config_fn}
+
+      Xebow.next_animation()
+      assert_receive ^unregister_message
+
+      Xebow.next_animation()
+      refute_receive ^unregister_message
+    end
+  end
```

Comment:
ℹ️ (optional)

FWIW, I don't see the `:unregister` atom in the body of this test, which would show how to use the behavior being described in the test description (documentation as code). It could help make the tests more clear if the important parts the test describes are in the body, with the boilerplate hidden away in helper functions.

Reply:
Well, since it's only the one test, it would be just as easy, and not too verbose, to define the function in the body of the test. It would make the usage a bit more obvious.

I keep going back and forth between wanting to define the tests first or the fixtures/helpers first. Both styles have their benefits....

Reply:
"YAGNI" says define the tests first, and create fixtures for code that starts turning into boilerplate. 😉

Reply:
Yeah, but I can envision situations where there are fixtures that you need to see to understand what a test is doing. I'd like to hope it's more rare, though, and also it seems like this hypothetical situation would indicate a bit of a code smell, possibly.